Asn1f2/Program.cs

```diff
@@ -411,64 +411,26 @@ public static int CheckSuccess(IEnumerable<string> args)
+            where L : Lexer where P : Parser
         {
-            var parsedInputFiles = new List<Tuple<ITree, string, IToken[]>>();
-
-            foreach (var inFileName in inputFiles)
-            {
-                ICharStream input = new ANTLRFileStream(inFileName);
-                asn1Lexer lexer = new asn1Lexer(input);
-                CommonTokenStream tokens = new CommonTokenStream(lexer);
-
-                List<IToken> tokenslst = new List<IToken>();
-                foreach (IToken token in tokens.GetTokens())
-                {
-                    tokenslst.Add(token);
-                }
-
-                asn1Parser parser = new asn1Parser(tokens);
-                asn1Parser.moduleDefinitions_return result = parser.moduleDefinitions();
-
-                ITree tree = (CommonTree)result.Tree;
-                CommonTreeNodeStream nodes = new CommonTreeNodeStream(tree);
-                nodes.TokenStream = tokens;
-                parsedInputFiles.Add(Tuple.Create(tree, inFileName, tokenslst.ToArray() ));
-            }
-
-            return parsedInputFiles;
+            var stream = new CommonTokenStream(lexer(new ANTLRFileStream(filename)));
+            var tokens = stream.GetTokens().Cast<IToken>().ToArray();
+            var tree = root(parser(stream));
+            return Tuple.Create(tree, filename, tokens);
         }
 
+        public static List<Tuple<ITree, string, IToken[]>> ParseAsn1InputFiles(IEnumerable<string> inputFiles)
+        {
+            return inputFiles.Select(f => Parse(f, fs => new asn1Lexer(fs), ts => new asn1Parser(ts), p => (CommonTree)p.moduleDefinitions().Tree)).ToList();
+        }
+
         public static List<Tuple<ITree, string, IToken[]>> ParseAcnInputFiles(IEnumerable<string> inputFiles)
         {
-            var parsedInputFiles = new List<Tuple<ITree, string, IToken[]>>();
-
-            foreach (var inFileName in inputFiles)
-            {
-                ICharStream input = new ANTLRFileStream(inFileName);
-                acnLexer lexer = new acnLexer(input);
-                CommonTokenStream tokens = new CommonTokenStream(lexer);
-
-                List<IToken> tokenslst = new List<IToken>();
-                foreach (IToken token in tokens.GetTokens())
-                {
-                    tokenslst.Add(token);
-                }
-
-                acnParser parser = new acnParser(tokens);
-                acnParser.moduleDefinitions_return result = parser.moduleDefinitions();
-
-                ITree tree = (CommonTree)result.Tree;
-                CommonTreeNodeStream nodes = new CommonTreeNodeStream(tree);
-                nodes.TokenStream = tokens;
-                parsedInputFiles.Add(Tuple.Create(tree, inFileName, tokenslst.ToArray()));
-            }
-
-            return parsedInputFiles;
+            return inputFiles.Select(f => Parse(f, fs => new acnLexer(fs), ts => new acnParser(ts), p => (CommonTree)p.moduleDefinitions().Tree)).ToList();

```

Comment:
Hi Alexandre.

It's clear there's a lot of repetition in that C# "gateway" code - I imagine George wrote it before he immersed himself in functional style (and F# in particular). The violation of DRY in some parts of the compiler (including Costas's testing Python script) is something I really wish I had time to dive in and fix - but my workload is not allowing me to do so. 

At least CircleCI shows that your patch doesn't seem to break any tests. If it were up to me, I'd merge it; but since I am not the author of this code I better let George decide.